protocol/auditlog.go

```diff
@@ -0,0 +1,202 @@
+// the first time it observes an STR for that directory.
+func NewAuditLog() ConiksAuditLog {
+	l := make(map[string]*directoryHistory)
+	return l
+}
+
+// IsKnownDirectory checks to see if an entry for the directory
+// (indexed by the hash of its initial STR dirInitHash) exists
+// in the audit log l. IsKnownDirectory() does not
+// validate the entries themselves. It returns true if an entry exists,
+// and false otherwise.
+func (l ConiksAuditLog) IsKnownDirectory(dirInitHash string) bool {
+	h := l[dirInitHash]
+	if h != nil {
+		return true
+	}
+	return false
+}
+
+// Insert creates a new directory history for the key directory addr,
+// verifies the consistency of the STR history so far, and inserts it
+// into the audit log l if the checks pass.
+// The directory history is initialized with the key directory's
+// signing key signKey, and a list of STRs representing the
+// directory's STR history so far (as a map of epochs to STRs).
+// Insert() returns an ErrAuditLog if the auditor attempts to create
+// a new history for a known directory, an ErrMalformedDirectoryMessage
+// if oldSTRs is malformed, a CheckBadSignature or CheckBadSTR if there
+// is an inconsistency in the history given in hist, and nil otherwise.
+// Insert() only creates the initial entry in the log for addr. Use Update()
+// to insert newly observed STRs for addr in subsequent epochs.
+// FIXME: pass Response message as param
+// masomel: will probably want to write a more generic function
+// for "catching up" on a history in case an auditor misses epochs
+func (l ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
+	hist map[uint64]*DirSTR) error {
+
+	// make sure we're getting an initial STR at the very least
+	if len(hist) < 1 && hist[0].Epoch != 0 {
+		return ErrMalformedDirectoryMessage
+	}
+
+	// compute the hash of the initial STR
+	dirInitHash := ComputeDirectoryIdentity(hist[0])
+
+	// error if we want to create a new entry for a directory
+	// we already know
+	if l.IsKnownDirectory(dirInitHash) {
+		return ErrAuditLog
+	}
+
+	// create the new directory history
+	h := newDirectoryHistory(addr, signKey, hist[0])
+
+	// add each STR into the history
+	// start at 1 since we've inserted the initial STR above
+	startEp := uint64(1)
+	endEp := uint64(len(hist))
+
+	// This loop automatically catches if hist is malformed
+	// (i.e. hist is missing an epoch between 0 and the latest given)
+	for ep := startEp; ep < endEp; ep++ {
+		str := hist[ep]
+		if str == nil {
+			return ErrMalformedDirectoryMessage
+		}
+
+		// verify the consistency of each new STR before inserting
+		// into the audit log
+		err := verifySTRConsistency(signKey, h.snapshots[ep-1], str)
+
+		if err != nil {
+			return err
+		}
+
+		h.snapshots[ep] = str
+	}
+
+	// Make sure to update the latestSTR
+	// in this particular call, the latestSTR has already been
+	// inserted into the snapshots map in the loop above
+	h.updateLatestSTR(hist[endEp-1])
```

Comment:
I think it would be cleared if you moved this to the last step of the loop, so that rather than

```
h.snapshots[ep] = str
```

you have,

```
h.updateLatestSTR(str)
```

then you won't need the clarifying comment, and you won't be doing it redundantly when `endEp == 1`.  It does mean a few unnecessary pointer updates, but I think it's worth it.